File: cs313_main/src/PeerClient.java
import java.net.*;

public class PeerClient {

  public static void main(String[] args) {

    System.out.println("""
        ██████╗  ██████╗ ██████╗  ██████╗███████╗██╗██████╗
        ██╔══██╗██╔═══██╗██╔══██╗██╔════╝██╔════╝██║██╔══██╗
        ██║  ██║██║   ██║██████╔╝██║     ███████╗██║██║  ██║
        ██║  ██║██║   ██║██╔══██╗██║     ╚════██║██║██║  ██║
        ██████╔╝╚██████╔╝██║  ██║╚██████╗███████║██║██████╔╝
        ╚═════╝  ╚═════╝ ╚═╝  ╚═╝ ╚═════╝╚══════╝╚═╝╚═════╝
                """);
    System.out.println("You are now experiencing DORCSID in PEER MODE");

    // Determine whether to listen or connect based on the nuber of command line
    // arguments given
    if (args.length >= 2) {
      String hostname = args[0];
      int portNo = Integer.parseInt(args[1]);

      System.out.println("Attempting to connect to client '" + hostname + "' on port " + portNo + "...");
      try (Socket socket = new Socket(hostname, portNo)) {
        System.out.println("Connection established! Enjoy your chat!");
        chatLoop(socket);
      } catch (Exception e) {
        e.printStackTrace();
      }
    } else if (args.length == 1) {
      int portNo = Integer.parseInt(args[0]);

      try (ServerSocket serverSocket = new ServerSocket(portNo)) {
        System.out.println("Listening for connection on port " + portNo + "...");
        try (Socket socket = serverSocket.accept()) {
          System.out.println("Connection established! Enjoy your chat!");
          chatLoop(socket);
        }
      } catch (Exception e) {
        e.printStackTrace();
      }
    }

  }

  // Sets up threads and runs the primary chat loop
  public static void chatLoop(Socket socket) {
    InputRunner inputRunner = new InputRunner(socket);
    OutputRunner outputRunner = new OutputRunner(socket);

    Thread inputThread = new Thread(inputRunner);
    Thread outputThread = new Thread(outputRunner);

    inputThread.start();
    outputThread.start();

    while (inputThread.isAlive() && outputThread.isAlive()) {
    }
  }

}

File: cs313_main/PeerClient.py
from socket import socket, AF_INET, SOCK_STREAM
import threading
import sys


def inputHandler(socket):
    while True:
        buffer = b""
        byte = b""
        while byte != b"\n":
            byte = socket.recv(1)
            if not byte:
                break
            buffer += byte
        print("\033[K", end="")
        print("\rDROCSID user> " + buffer.decode("utf-8"), end="")
        print("\033[Kyou> ", end="", flush=True)


def outputHandler(socket):
    while True:
        message = input("you> ")
        message += "\n"
        socket.send(message.encode("utf-8"))


print(
    """
██████╗ ██████╗  ██████╗  ██████╗███████╗██╗██████╗
██╔══██╗██╔══██╗██╔═══██╗██╔════╝██╔════╝██║██╔══██╗
██║  ██║██████╔╝██║   ██║██║     ███████╗██║██║  ██║
██║  ██║██╔══██╗██║   ██║██║     ╚════██║██║██║  ██║
██████╔╝██║  ██║╚██████╔╝╚██████╗███████║██║██████╔╝
╚═════╝ ╚═╝  ╚═╝ ╚═════╝  ╚═════╝╚══════╝╚═╝╚═════╝
    """
)

print("You are now experiencing DROCSID in PYTHON PEER MODE")


if len(sys.argv) >= 3:
    hostname = sys.argv[1]
    port_no = int(sys.argv[2])
    socket = socket(AF_INET, SOCK_STREAM)
    socket.connect((hostname, port_no))
elif len(sys.argv) == 2:
    port_no = int(sys.argv[1])
    server_socket = socket(AF_INET, SOCK_STREAM)
    server_socket.bind(("", port_no))
    server_socket.listen()
    socket, addr = server_socket.accept()

inputThread = threading.Thread(target=inputHandler, args=(socket,))
outputThread = threading.Thread(target=outputHandler, args=(socket,))

inputThread.start()
outputThread.start()

while inputThread.is_alive() and outputThread.is_alive():
    pass
